### ISSUE TITLE
COMPRESSION_ERROR occurs when the SETTINGS_HEADER_TABLE_SIZE value of the HTTP2 client is less than 4096.

### ISSUE THREAD

**Issue by jinsubsim** [haproxy/haproxy#1498]
### Detailed Description of the Problem

There is an issue that the HTTP2 Client cannot process the response transmitted by HAProxy when the HTTP2 Client sets the SETTINGS_HEADER_TABLE_SIZE value less then 4096. 
``` 
https://datatracker.ietf.org/doc/html/rfc7540#section-6.5.2

6.5.2.  Defined SETTINGS Parameters

   The following parameters are defined:

   SETTINGS_HEADER_TABLE_SIZE (0x1):  Allows the sender to inform the
      remote endpoint of the maximum size of the header compression
      table used to decode header blocks, in octets.  The encoder can
      select any size equal to or less than this value by using
      signaling specific to the header compression format inside a
      header block (see [COMPRESSION]).  The initial value is 4,096
      octets.
```

### Expected Behavior

There is no constraint that SETTINGS_HEADER_TABLE_SIZE should be set above 4096 even in the RFC7540 (Hypertext Transfer Protocol Version) and RFC7541 (HPACK: Header Compression for HTTP/2).

Therefore, I think that even if the value of SETTINGS_HEADER_TABLE_SIZE is less than 4096, it should be interworked with HAProxy normally.

### Steps to Reproduce the Behavior

1. Install nghttp application.
```
$ sudo apt install nghttp2-client
```
2. Run haproxy and backend server
3. Send GET request to the haproxy by nghttp and then check the result with "error_code=COMPRESSION_ERROR(0x09)".
```
$ nghttp -v --header-table-size=0 http://localhost:8080 
[  0.000] Connected
[  0.000] send SETTINGS frame <length=18, flags=0x00, stream_id=0>
          (niv=3)
          [SETTINGS_MAX_CONCURRENT_STREAMS(0x03):100]
          [SETTINGS_INITIAL_WINDOW_SIZE(0x04):65535]
          [SETTINGS_HEADER_TABLE_SIZE(0x01):0]
[  0.000] send PRIORITY frame <length=5, flags=0x00, stream_id=3>
          (dep_stream_id=0, weight=201, exclusive=0)
[  0.000] send PRIORITY frame <length=5, flags=0x00, stream_id=5>
          (dep_stream_id=0, weight=101, exclusive=0)
[  0.000] send PRIORITY frame <length=5, flags=0x00, stream_id=7>
          (dep_stream_id=0, weight=1, exclusive=0)
[  0.000] send PRIORITY frame <length=5, flags=0x00, stream_id=9>
          (dep_stream_id=7, weight=1, exclusive=0)
[  0.000] send PRIORITY frame <length=5, flags=0x00, stream_id=11>
          (dep_stream_id=3, weight=1, exclusive=0)
[  0.000] send HEADERS frame <length=38, flags=0x25, stream_id=13>
          ; END_STREAM | END_HEADERS | PRIORITY
          (padlen=0, dep_stream_id=11, weight=16, exclusive=0)
          ; Open new stream
          :method: GET
          :path: /
          :scheme: http
          :authority: localhost:8080
          accept: */*
          accept-encoding: gzip, deflate
          user-agent: nghttp2/1.40.0
[  0.000] recv SETTINGS frame <length=12, flags=0x00, stream_id=0>
          (niv=2)
          [SETTINGS_ENABLE_CONNECT_PROTOCOL(0x08):1]
          [SETTINGS_MAX_CONCURRENT_STREAMS(0x03):100]
[  0.000] recv SETTINGS frame <length=0, flags=0x01, stream_id=0>
          ; ACK
          (niv=0)
[  0.000] send SETTINGS frame <length=0, flags=0x01, stream_id=0>
          ; ACK
          (niv=0)
[  0.001] send GOAWAY frame <length=8, flags=0x00, stream_id=0>
          (last_stream_id=0, error_code=COMPRESSION_ERROR(0x09), opaque_data(0)=[])
Some requests were not processed. total=1, processed=0
```

### Do you have any idea what may have caused this?

The SETTINGS_HEADER_TABLE_SIZE value of the client's SETTINGS frame is not handled by h2c_handle_settings() function in mux_h2.c.

### Do you have an idea how to solve the issue?

According to Section 6.3 of RFC7541, it is stated that Dynamic Table Size can be updated.
```
https://datatracker.ietf.org/doc/html/rfc7541#section-6.3

6.3.  Dynamic Table Size Update

   A dynamic table size update signals a change to the size of the
   dynamic table.

     0   1   2   3   4   5   6   7
   +---+---+---+---+---+---+---+---+
   | 0 | 0 | 1 |   Max size (5+)   |
   +---+---------------------------+

               Figure 12: Maximum Dynamic Table Size Change

   A dynamic table size update starts with the '001' 3-bit pattern,
   followed by the new maximum size, represented as an integer with a
   5-bit prefix (see Section 5.1).

   The new maximum size MUST be lower than or equal to the limit
   determined by the protocol using HPACK.  A value that exceeds this
   limit MUST be treated as a decoding error.  In HTTP/2, this limit is
   the last value of the SETTINGS_HEADER_TABLE_SIZE parameter (see
   Section 6.5.2 of [HTTP2]) received from the decoder and acknowledged
   by the encoder (see Section 6.5.3 of [HTTP2]).

   Reducing the maximum size of the dynamic table can cause entries to
   be evicted (see Section 4.3).
```

So after modifying code as below, I did the same test and got normal result.

```
$ git diff
diff --git a/src/mux_h2.c b/src/mux_h2.c
index dfe0b37..71efc95 100644
--- a/src/mux_h2.c
+++ b/src/mux_h2.c
@@ -105,6 +105,7 @@ struct h2c {
        /* states for the demux direction */
        struct hpack_dht *ddht; /* demux dynamic header table */
        struct buffer dbuf;    /* demux buffer */
+       uint32_t hts; /* header table size of client */
 
        int32_t dsi; /* demux stream ID (<0 = idle) */
        int32_t dfl; /* demux frame length (if dsi >= 0) */
@@ -993,6 +994,7 @@ static int h2_init(struct connection *conn, struct proxy *prx, struct session *s
        h2c->msi = -1;
 
        h2c->last_sid = -1;
+       h2c->hts = 4096;
 
        br_init(h2c->mbuf, sizeof(h2c->mbuf) / sizeof(h2c->mbuf[0]));
        h2c->miw = 65535; /* mux initial window size */
@@ -2190,6 +2192,13 @@ static int h2c_handle_settings(struct h2c *h2c)
                int32_t  arg  = h2_get_n32(&h2c->dbuf, offset + 2);
 
                switch (type) {
+        case H2_SETTINGS_HEADER_TABLE_SIZE:
+            if (arg < 0 || arg > 65536) {
+               TRACE_ERROR("HEADER_TABLESIZE out of range", H2_EV_RX_FRAME|H2_EV_RX_SETTINGS, h2c->conn);
+                goto fail;
+            }
+                       h2c->hts = arg;
+            break;
                case H2_SETTINGS_INITIAL_WINDOW_SIZE:
                        /* we need to update all existing streams with the
                         * difference from the previous iws.
@@ -5112,6 +5121,12 @@ static size_t h2s_frt_make_resp_headers(struct h2s *h2s, struct htx *htx)
        write_n32(outbuf.area + 5, h2s->id); // 4 bytes
        outbuf.data = 9;
 
+       if (h2c->hts < 4096)
+       {
+               outbuf.area[outbuf.data] = '\x20';
+               outbuf.data += 1;
+       }
+
        /* encode status, which necessarily is the first one */
        if (!hpack_encode_int_status(&outbuf, h2s->status)) {
                if (b_space_wraps(mbuf))
```

Normal Result : There is no COMPRESSION_ERROR (0x09).
```
$ nghttp -v --header-table-size=0 http://localhost:8080 
[  0.001] Connected
[  0.001] send SETTINGS frame <length=18, flags=0x00, stream_id=0>
          (niv=3)
          [SETTINGS_MAX_CONCURRENT_STREAMS(0x03):100]
          [SETTINGS_INITIAL_WINDOW_SIZE(0x04):65535]
          [SETTINGS_HEADER_TABLE_SIZE(0x01):0]
[  0.001] send PRIORITY frame <length=5, flags=0x00, stream_id=3>
          (dep_stream_id=0, weight=201, exclusive=0)
[  0.001] send PRIORITY frame <length=5, flags=0x00, stream_id=5>
          (dep_stream_id=0, weight=101, exclusive=0)
[  0.001] send PRIORITY frame <length=5, flags=0x00, stream_id=7>
          (dep_stream_id=0, weight=1, exclusive=0)
[  0.001] send PRIORITY frame <length=5, flags=0x00, stream_id=9>
          (dep_stream_id=7, weight=1, exclusive=0)
[  0.001] send PRIORITY frame <length=5, flags=0x00, stream_id=11>
          (dep_stream_id=3, weight=1, exclusive=0)
[  0.001] send HEADERS frame <length=38, flags=0x25, stream_id=13>
          ; END_STREAM | END_HEADERS | PRIORITY
          (padlen=0, dep_stream_id=11, weight=16, exclusive=0)
          ; Open new stream
          :method: GET
          :path: /
          :scheme: http
          :authority: localhost:8080
          accept: */*
          accept-encoding: gzip, deflate
          user-agent: nghttp2/1.40.0
[  0.001] recv SETTINGS frame <length=12, flags=0x00, stream_id=0>
          (niv=2)
          [SETTINGS_ENABLE_CONNECT_PROTOCOL(0x08):1]
          [SETTINGS_MAX_CONCURRENT_STREAMS(0x03):100]
[  0.001] recv SETTINGS frame <length=0, flags=0x01, stream_id=0>
          ; ACK
          (niv=0)
[  0.001] send SETTINGS frame <length=0, flags=0x01, stream_id=0>
          ; ACK
          (niv=0)
[  0.005] recv (stream_id=13) :status: 200
[  0.005] recv (stream_id=13) content-length: 35
[  0.005] recv (stream_id=13) custom-header: header table size test
[  0.005] recv (stream_id=13) server: nghttp2_server
[  0.005] recv HEADERS frame <length=60, flags=0x04, stream_id=13>
          ; END_HEADERS
          (padlen=0)
          ; First response header
{
        "message": "HELLO_ACK"
}
[  0.005] recv DATA frame <length=35, flags=0x01, stream_id=13>
          ; END_STREAM
[  0.005] send GOAWAY frame <length=8, flags=0x00, stream_id=0>
          (last_stream_id=0, error_code=NO_ERROR(0x00), opaque_data(0)=[])
```

### What is your configuration?

```haproxy
$ cat haproxy.cfg 
global 
   nbthread 2
   log /dev/log local0 info
   log /dev/log local0 notice
   log /dev/log local0 debug
   user        root 
   group       root

defaults
   log global
   log     127.0.0.1       local0
   option  httplog
   timeout connect 10s
   timeout client 40s
   timeout server 1h
   maxconn 30000
   mode http

frontend http-in
   bind *:8080 proto h2
   default_backend   servers

backend servers
   balance roundrobin
   server srv1 127.0.0.1:23232 proto h2
```

### Output of `haproxy -vv`

```plain
$ ./haproxy -vv
HAProxy version 2.4.7-b5e51a5 2021/10/04 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2026.
Known bugs: http://www.haproxy.org/bugs/bugs-2.4.7.html
Running on: Linux 5.11.0-40-generic #44~20.04.2-Ubuntu SMP Tue Oct 26 18:07:44 UTC 2021 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -Wall -Wextra -Wdeclaration-after-statement -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = 
  DEBUG   = 

Feature list : +EPOLL -KQUEUE +NETFILTER -PCRE -PCRE_JIT -PCRE2 -PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED +BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H +GETADDRINFO -OPENSSL -LUA +FUTEX +ACCEPT4 -CLOSEFROM -ZLIB +SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL -SYSTEMD -OBSOLETE_LINKER +PRCTL -PROCCTL +THREAD_DUMP -EVPORTS -OT -QUIC -PROMEX -MEMORY_PROFILING

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=8).
Built with network namespace support.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built without PCRE or PCRE2 support (using libc's regex instead)
Encrypted password support via crypt(3): yes
Built with gcc compiler version 9.3.0

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTTP       side=FE|BE     mux=H2       flags=HTX|CLEAN_ABRT|HOL_RISK|NO_UPG
            fcgi : mode=HTTP       side=BE        mux=FCGI     flags=HTX|HOL_RISK|NO_UPG
       <default> : mode=HTTP       side=FE|BE     mux=H1       flags=HTX
              h1 : mode=HTTP       side=FE|BE     mux=H1       flags=HTX|NO_UPG
       <default> : mode=TCP        side=FE|BE     mux=PASS     flags=
            none : mode=TCP        side=FE|BE     mux=PASS     flags=NO_UPG

Available services : none

Available filters :
	[SPOE] spoe
	[CACHE] cache
	[FCGI] fcgi-app
	[COMP] compression
	[TRACE] trace
```

**Comment by wtarreau** [comment]
First, many thanks for your detailed analysis.

I think we're facing an interoperability issue here, because the way I read RFC7541 we do not have to send an update here since the table size defaults to the one advertised by the peers' settings and we're not changing it:

```
   Protocols that use HPACK determine the maximum size that the encoder
   is permitted to use for the dynamic table.  In HTTP/2, this value is
   determined by the SETTINGS_HEADER_TABLE_SIZE setting (see
   Section 6.5.2 of [HTTP2]).
```

Thus nghttp indicates that its table size is smaller than 4kB. Then:

```
   An encoder can choose to use less capacity than this maximum size
   (see Section 6.3), but the chosen size MUST stay lower than or equal
   to the maximum set by the protocol.
```

Here we're not choosing to use a lower size, we just conform to the advertised size.

```
   A change in the maximum size of the dynamic table is signaled via a
   dynamic table size update (see Section 6.3).  This dynamic table size
   update MUST occur at the beginning of the first header block
   following the change to the dynamic table size.  In HTTP/2, this
   follows a settings acknowledgment (see Section 6.5.3 of [HTTP2]).
```

And here it justifies your change of course. I wouldn't necessarily be opposed to always advertising a zero-sized  dictionary since we don't use it, but I think that here nghttp and us are possibly having a different interpretation of the spec that needs to be sorted out.

**Comment by jinsubsim** [comment]
First of all, thanks for the reply.

It seems to be possible to send the Dynamic Table update message only once for the first time instead of sending it every time.
My intention is not to use the Dynamic Table at all when the Client's SETTINGS_HEADER_TABLE_SIZE is less than 4096.

Could you suggest another way to solve this situation?

**Comment by wtarreau** [comment]
We *never* use the dynamic table when sending, regardless of the advertised size, which is also why we don't even do anything with the size in the settings. That's why I suspect that the only issue is a different interpretation of the spec. I'll possibly bring the question to the HTTP-WG after discussing it with Tatsuhiro.

Have you checked if when sending this request to another server there's a smaller table advertised by the server ? Really I'm not against doing it on every connection (it's only one byte, we can afford that). I just want to be certain it is needed because my reading of the spec tells me otherwise.

**Comment by jinsubsim** [comment]
I did the same tests in multiple environments.

1) nghttp client (SETTINGS_HEADER_TABLE_SIZE: 0) <---> nghttp2 server
- Result: OK (Send a response including dynamic table update (0x20))
```
$ nghttp -v --header-table-size=0 http://localhost:23232 
[  0.000] Connected
[  0.000] send SETTINGS frame <length=18, flags=0x00, stream_id=0>
          (niv=3)
          [SETTINGS_MAX_CONCURRENT_STREAMS(0x03):100]
          [SETTINGS_INITIAL_WINDOW_SIZE(0x04):65535]
          [SETTINGS_HEADER_TABLE_SIZE(0x01):0]
[  0.000] send PRIORITY frame <length=5, flags=0x00, stream_id=3>
          (dep_stream_id=0, weight=201, exclusive=0)
[  0.000] send PRIORITY frame <length=5, flags=0x00, stream_id=5>
          (dep_stream_id=0, weight=101, exclusive=0)
[  0.000] send PRIORITY frame <length=5, flags=0x00, stream_id=7>
          (dep_stream_id=0, weight=1, exclusive=0)
[  0.000] send PRIORITY frame <length=5, flags=0x00, stream_id=9>
          (dep_stream_id=7, weight=1, exclusive=0)
[  0.000] send PRIORITY frame <length=5, flags=0x00, stream_id=11>
          (dep_stream_id=3, weight=1, exclusive=0)
[  0.000] send HEADERS frame <length=39, flags=0x25, stream_id=13>
          ; END_STREAM | END_HEADERS | PRIORITY
          (padlen=0, dep_stream_id=11, weight=16, exclusive=0)
          ; Open new stream
          :method: GET
          :path: /
          :scheme: http
          :authority: localhost:23232
          accept: */*
          accept-encoding: gzip, deflate
          user-agent: nghttp2/1.40.0
[  0.001] recv SETTINGS frame <length=12, flags=0x00, stream_id=0>
          (niv=2)
          [SETTINGS_MAX_CONCURRENT_STREAMS(0x03):180000]
          [SETTINGS_INITIAL_WINDOW_SIZE(0x04):6291456]
[  0.001] recv SETTINGS frame <length=0, flags=0x01, stream_id=0>
          ; ACK
          (niv=0)
[  0.001] recv WINDOW_UPDATE frame <length=4, flags=0x00, stream_id=0>
          (window_size_increment=6225921)
[  0.001] send SETTINGS frame <length=0, flags=0x01, stream_id=0>
          ; ACK
          (niv=0)
[  0.001] recv (stream_id=13) :status: 200
[  0.001] recv (stream_id=13) content-length: 35
[  0.001] recv (stream_id=13) custom-header: header table size test
[  0.001] recv (stream_id=13) server: nghttp2_server
[  0.001] recv HEADERS frame <length=47, flags=0x04, stream_id=13>
          ; END_HEADERS
          (padlen=0)
          ; First response header
{
        "message": "HELLO_ACK"
}
[  0.001] recv DATA frame <length=35, flags=0x01, stream_id=13>
          ; END_STREAM
[  0.001] send GOAWAY frame <length=8, flags=0x00, stream_id=0>
          (last_stream_id=0, error_code=NO_ERROR(0x00), opaque_data(0)=[])
```


2) nghttp client (SETTINGS_HEADER_TABLE_SIZE: 0) <---> envoy server
- Result: OK (Send a response including dynamic table update (0x20))
```
$ nghttp -v --header-table-size=0 http://localhost:31380/hello
[  0.000] Connected
[  0.001] send SETTINGS frame <length=18, flags=0x00, stream_id=0>
          (niv=3)
          [SETTINGS_MAX_CONCURRENT_STREAMS(0x03):100]
          [SETTINGS_INITIAL_WINDOW_SIZE(0x04):65535]
          [SETTINGS_HEADER_TABLE_SIZE(0x01):0]
[  0.001] send PRIORITY frame <length=5, flags=0x00, stream_id=3>
          (dep_stream_id=0, weight=201, exclusive=0)
[  0.001] send PRIORITY frame <length=5, flags=0x00, stream_id=5>
          (dep_stream_id=0, weight=101, exclusive=0)
[  0.001] send PRIORITY frame <length=5, flags=0x00, stream_id=7>
          (dep_stream_id=0, weight=1, exclusive=0)
[  0.001] send PRIORITY frame <length=5, flags=0x00, stream_id=9>
          (dep_stream_id=7, weight=1, exclusive=0)
[  0.001] send PRIORITY frame <length=5, flags=0x00, stream_id=11>
          (dep_stream_id=3, weight=1, exclusive=0)
[  0.001] send HEADERS frame <length=48, flags=0x25, stream_id=13>
          ; END_STREAM | END_HEADERS | PRIORITY
          (padlen=0, dep_stream_id=11, weight=16, exclusive=0)
          ; Open new stream
          :method: GET
          :path: /hello
          :scheme: http
          :authority: 1localhost:31380
          accept: */*
          accept-encoding: gzip, deflate
          user-agent: nghttp2/1.40.0
[  0.002] recv SETTINGS frame <length=6, flags=0x00, stream_id=0>
          (niv=1)
          [SETTINGS_INITIAL_WINDOW_SIZE(0x04):268435456]
[  0.002] recv SETTINGS frame <length=0, flags=0x01, stream_id=0>
          ; ACK
          (niv=0)
[  0.002] recv WINDOW_UPDATE frame <length=4, flags=0x00, stream_id=0>
          (window_size_increment=268369921)
[  0.002] recv (stream_id=13) :status: 404
[  0.002] recv (stream_id=13) ok: KO!!!!
[  0.002] recv (stream_id=13) date: Wed, 22 Dec 2021 13:39:09 GMT
[  0.002] recv (stream_id=13) server: istio-envoy
[  0.002] recv HEADERS frame <length=49, flags=0x05, stream_id=13>
          ; END_STREAM | END_HEADERS
          (padlen=0)
          ; First response header
[  0.002] send GOAWAY frame <length=8, flags=0x00, stream_id=0>
          (last_stream_id=0, error_code=NO_ERROR(0x00), opaque_data(0)=[])

```

3) nghttp client (SETTINGS_HEADER_TABLE_SIZE: 0) <---> nginx server
- Result: OK (Send a response including dynamic table update (0x20))
```
$ nghttp -v --header-table-size=0 http://localhost:8080 
[  0.000] Connected
[  0.000] send SETTINGS frame <length=18, flags=0x00, stream_id=0>
          (niv=3)
          [SETTINGS_MAX_CONCURRENT_STREAMS(0x03):100]
          [SETTINGS_INITIAL_WINDOW_SIZE(0x04):65535]
          [SETTINGS_HEADER_TABLE_SIZE(0x01):0]
[  0.000] send PRIORITY frame <length=5, flags=0x00, stream_id=3>
          (dep_stream_id=0, weight=201, exclusive=0)
[  0.000] send PRIORITY frame <length=5, flags=0x00, stream_id=5>
          (dep_stream_id=0, weight=101, exclusive=0)
[  0.000] send PRIORITY frame <length=5, flags=0x00, stream_id=7>
          (dep_stream_id=0, weight=1, exclusive=0)
[  0.000] send PRIORITY frame <length=5, flags=0x00, stream_id=9>
          (dep_stream_id=7, weight=1, exclusive=0)
[  0.000] send PRIORITY frame <length=5, flags=0x00, stream_id=11>
          (dep_stream_id=3, weight=1, exclusive=0)
[  0.000] send HEADERS frame <length=38, flags=0x25, stream_id=13>
          ; END_STREAM | END_HEADERS | PRIORITY
          (padlen=0, dep_stream_id=11, weight=16, exclusive=0)
          ; Open new stream
          :method: GET
          :path: /
          :scheme: http
          :authority: localhost:8080
          accept: */*
          accept-encoding: gzip, deflate
          user-agent: nghttp2/1.40.0
[  0.000] recv SETTINGS frame <length=18, flags=0x00, stream_id=0>
          (niv=3)
          [SETTINGS_MAX_CONCURRENT_STREAMS(0x03):128]
          [SETTINGS_INITIAL_WINDOW_SIZE(0x04):65536]
          [SETTINGS_MAX_FRAME_SIZE(0x05):16777215]
[  0.000] recv WINDOW_UPDATE frame <length=4, flags=0x00, stream_id=0>
          (window_size_increment=2147418112)
[  0.000] recv SETTINGS frame <length=0, flags=0x01, stream_id=0>
          ; ACK
          (niv=0)
[  0.000] recv (stream_id=13) :status: 200
[  0.000] recv (stream_id=13) server: nginx/1.18.0 (Ubuntu)
[  0.000] recv (stream_id=13) date: Wed, 22 Dec 2021 13:36:22 GMT
[  0.000] recv (stream_id=13) content-type: text/html
[  0.000] recv (stream_id=13) last-modified: Wed, 22 Dec 2021 13:12:39 GMT
[  0.000] recv (stream_id=13) etag: W/"61c32447-264"
[  0.000] recv (stream_id=13) content-encoding: gzip
[  0.000] recv HEADERS frame <length=114, flags=0x04, stream_id=13>
          ; END_HEADERS
          (padlen=0)
          ; First response header
<!DOCTYPE html>
<html>
<head>
<title>Welcome to nginx!</title>
<style>
    body {
        width: 35em;
        margin: 0 auto;
        font-family: Tahoma, Verdana, Arial, sans-serif;
    }
</style>
</head>
<body>
<h1>Welcome to nginx!</h1>
<p>If you see this page, the nginx web server is successfully installed and
working. Further configuration is required.</p>

<p>For online documentation and support please refer to
<a href="http://nginx.org/">nginx.org</a>.<br/>
Commercial support is available at
<a href="http://nginx.com/">nginx.com</a>.</p>

<p><em>Thank you for using nginx.</em></p>
</body>
</html>
[  0.000] recv DATA frame <length=384, flags=0x01, stream_id=13>
          ; END_STREAM
[  0.000] send GOAWAY frame <length=8, flags=0x00, stream_id=0>
          (last_stream_id=0, error_code=NO_ERROR(0x00), opaque_data(0)=[])
```

**Comment by wtarreau** [comment]
Thank you. It's something we can do but it doesn't correspond to what the spec says, and Tastuhiro pointed to an old conversation where other implementers (including one of the spec's authors) as well suggested that this was not necessary.

But that's the principle of interoperability, we need to find the best way forward when a spec isn't very clear, and sometimes usage becomes the new spec. Ideally I think that future versions of nghttp2 ought not to expect that initial byte and that haproxy ought to emit it so that we improve the coverage between older and newer versions.

**Comment by jinsubsim** [comment]
I understand that the specs are unclear.

Does your answer mean that you will add that HAProxy emits 0x20?

**Comment by wtarreau** [comment]
It's one of the possible outcome. The question was raised within the HTTP working group here: https://lists.w3.org/Archives/Public/ietf-http-wg/2021OctDec/0235.html

We'll see how to sort this out. As I said I'm not against doing it provided that there is consensus around this behavior which can be one possible interpretation of the spec but not necessarily the only one. I just don't want to add random bytes that are difficult to justify "just because it seems to work better this way".

**Comment by jinsubsim** [comment]
Thanks for your update. :)

**Comment by wtarreau** [comment]
Some progress is being made on the thread above and there's a PR to clarify this point in the imminent revision of the H2 spec. With it I'm fine with adding that 0x20 byte at the beginning, I'll look at how to do it unconditionally since we never use the dictionary anyway, but we just want to send it once.

**Comment by wtarreau** [comment]
Update: for now I'm postponing that. It breaks vtest, and checking a bunch of other sites shows me that a small majority of servers behave like haproxy (apache,varnish,h2o,litespeed). Thus we need to sort this out within the working group before making a move in any direction.

**Comment by jinsubsim** [comment]
Oh! I see.
Thanks for sharing good information.

**Comment by wtarreau** [comment]
It can take a bit more time, for a reason I do not understand, Tatsuhiro decided that *his* interpretation of the spec is the only correct one, that all servers that read it differently are broken, refuses further discussion, and locked the issue to make sure we don't discuss about it anymore :-(  Even if the spec is amended to go into that direction, it will not suddenly make servers that comply with the current one any less compliant and some smooth solutions will have to be found anyway :-/

**Comment by jinsubsim** [comment]
Hello. Tarreau!

Any updates?
It has been reflected in the spec, so please review it positively.



FYI: Below is a sample code modified to transmit dynamic table size update information only once.
```
$ git diff
diff --git a/src/mux_h2.c b/src/mux_h2.c
index dfe0b37..a16fc81 100644
--- a/src/mux_h2.c
+++ b/src/mux_h2.c
@@ -105,6 +105,7 @@ struct h2c {
        /* states for the demux direction */
        struct hpack_dht *ddht; /* demux dynamic header table */
        struct buffer dbuf;    /* demux buffer */
+       uint32_t hts; /* header table size of decorder */
 
        int32_t dsi; /* demux stream ID (<0 = idle) */
        int32_t dfl; /* demux frame length (if dsi >= 0) */
@@ -200,6 +201,7 @@ enum h2_ss {
 #define H2_SF_EXT_CONNECT_RCVD  0x00080000  // rfc 8441 an Extended CONNECT has been received and parsed
 
 #define H2_SF_TUNNEL_ABRT       0x00100000  // A tunnel attempt was aborted
+#define H2_SF_DTSIZE_SENT       0x00200000  // sent dynamic table size
 
 /* H2 stream descriptor, describing the stream as it appears in the H2C, and as
  * it is being processed in the internal HTTP representation (HTX).
@@ -993,6 +995,7 @@ static int h2_init(struct connection *conn, struct proxy *prx, struct session *s
        h2c->msi = -1;
 
        h2c->last_sid = -1;
+       h2c->hts = 4096;
 
        br_init(h2c->mbuf, sizeof(h2c->mbuf) / sizeof(h2c->mbuf[0]));
        h2c->miw = 65535; /* mux initial window size */
@@ -2190,6 +2193,18 @@ static int h2c_handle_settings(struct h2c *h2c)
                int32_t  arg  = h2_get_n32(&h2c->dbuf, offset + 2);
 
                switch (type) {
+               case H2_SETTINGS_HEADER_TABLE_SIZE:
+                       if (arg < 0 || arg > 65536) {
+                               TRACE_ERROR("HEADER_TABLESIZE out of range", H2_EV_RX_FRAME | H2_EV_RX_SETTINGS, h2c->conn);
+                               error = H2_ERR_PROTOCOL_ERROR;
+                               HA_ATOMIC_INC(&h2c->px_counters->conn_proto_err);
+                               goto fail;
+                       }
+                       if (arg < h2c->hts) {
+                               h2c->hts = arg;
+                               h2c->flags &= ~H2_SF_DTSIZE_SENT;
+                       }
+                       break;
                case H2_SETTINGS_INITIAL_WINDOW_SIZE:
                        /* we need to update all existing streams with the
                         * difference from the previous iws.
@@ -5112,6 +5127,11 @@ static size_t h2s_frt_make_resp_headers(struct h2s *h2s, struct htx *htx)
        write_n32(outbuf.area + 5, h2s->id); // 4 bytes
        outbuf.data = 9;
 
+       if (!(h2c->flags & H2_SF_DTSIZE_SENT)) { // RFC7541#4.2
+               outbuf.area[outbuf.data] = '\x20';
+               outbuf.data += 1;
+       }
+
        /* encode status, which necessarily is the first one */
        if (!hpack_encode_int_status(&outbuf, h2s->status)) {
                if (b_space_wraps(mbuf))
@@ -5205,6 +5225,9 @@ static size_t h2s_frt_make_resp_headers(struct h2s *h2s, struct htx *htx)
                        h2s_close(h2s);
        }
 
+       if (!(h2c->flags & H2_SF_DTSIZE_SENT))
+               h2c->flags |= H2_SF_DTSIZE_SENT;
+
        /* OK we could properly deliver the response */
  end:
        TRACE_LEAVE(H2_EV_TX_FRAME|H2_EV_TX_HDR, h2c->conn, h2s);
```

**Comment by wtarreau** [comment]
For now the change in the spec is still under discussion between the various participants. As other products are concerned as well, it's important not to rush a quick solution and to make sure the chosen one is reasonably acceptable everywhere and will not add more trouble. There are some absurdities in this situation, for example one agent might have prepared and buffered an HPACK message and might need to postpone an H2 SETTINGS ACK just to stay compliant. Overall this change is causing some layering violations that could have real impacts if poorly done.

**Comment by jinsubsim** [comment]
OK!
Thank you for your answer.

**Comment by wtarreau** [comment]
@jinsubsim I've prepared this patch which closely matches what the current PR suggests and which seems to be reaching consensus. It works fine with nghttp2 and doesn't break vtest anymore. Do not hesitate to test and make any comment. I'll merge it once the spec is updated.
[0001-MEDIUM-h2-hpack-emit-a-Dynamic-Table-Size-Update-aft.patch.txt](https://github.com/haproxy/haproxy/files/7863972/0001-MEDIUM-h2-hpack-emit-a-Dynamic-Table-Size-Update-aft.patch.txt)

**Comment by jinsubsim** [comment]
I applied your patch and tested it.
It works Good!

Thank you for your support.  :)

**Comment by wtarreau** [comment]
Perfect, it's OK for me as well here. I'll merge it soon I guess.

**Comment by wtarreau** [comment]
Now merged, It will get backported to 2.5 And 2.4 soon, and progressively to older versions.

**Comment by jinsubsim** [comment]
Great! Thanks a lot. :)

**Comment by capflam** [comment]
The fix will not be backported to older versions. I'm closing the issue now.
